### ISSUE TITLE
Implement a simple camelCase transform

### ISSUE THREAD

**Issue by magcius** [WebReflection/node-gtk#9]
To ease writing JS code, one of the proposals we have is to allow users to opt-in to a camelCaseifier type module, similar to how bluebird has `promisifyAll`. One would imagine that it could look like this:

``` js
var Gtk = gi.importNs('Gtk');
var ngc = require('node-gtk-camelcase');
ngc.camelCaseifyAll(Gtk);

var window = new Gtk.Window({ iconName: "my-application" });
window.setSizeRequest(800, 600);
window.showAll();

console.log(window.getTitle());
```

This would add camel-case equivalents to all functions and properties.

The hope is that by making this explicit, it becomes less like incomprehensible magic to the community, and it can help explain edge cases e.g. signals still being named `"notify::icon-name"` because our module does not mangle signal names automatically -- only functions and properties.

There's also plenty of other strange kinds and edge cases to tackle. One simple example is acronym based names. Obviously, we'll only be able to automatically transform to `createHttpServer` and not `createHTTPServer`, but I'm not sure if this is natural enough for JS users. I'm not even sure the JS community has a consistent standard for this (you can see this in the amazingly named `XMLHttpRequest`)

Anyway, let's build a simple, opt-in transform module to test it out.

**Comment by WebReflection** [comment]
We can create special cases without problems. The main issue I see is replacing Gtk.Window constructor in order to be able to intercept setup objects. It causes inconsistent results via `instanceof` . About signals, events in JS have always been lowercase plus you can have any string so I don't see any problem with signals names also because signals don't exist in node so it's like an API a part.

**Comment by WebReflection** [comment]
To be fair, lowercase with few exceptions such DOMContentLoaded....but that's another story, every other older or newer event name has been made consistent. Then again, everyone can write `onStuff::whatever`

**Comment by magcius** [comment]
As for constructor hashes, I would actually probably write a hook the C constructor can call to normalize the hash. Before we pass to the underlying constructor, we normalize the values.

Actually, since `External` objects are values, I wonder if you can propagate them around in JS code as opaque values. That might give us the opportunity to remove a lot more of the complex C++ code in `gobject.cc` and implement the constructors in plain JS code.

**Comment by WebReflection** [comment]
Unfortunately V8 is not even close to implement proper `Proxy` which could basically make everything interceptable in JSland delegating to fewer wrapped functions. However, if there is a reliable way to know all properties/methods/statics per each GObject at runtime we could lazily build on demand every constructor. In GJS this is not currently possible because the API is discovered at runtime but not exposed in JS land via `Object.getOwnPropertyNames(generic)` 

Not sure how much I've actually clarified or if I've add confusion... Or maybe I misunderstood your question.

**Comment by benwaffle** [comment]
would using a polyfill for Proxy be a bad idea?

**Comment by magcius** [comment]
`Proxy` is unrelated, we wouldn't use it in our transform module. So I believe you misunderstood me.

**Comment by WebReflection** [comment]
@magcius fair enough, mind explaining with a code example what do you have in your mind?

@benwaffle impossible to polyfill ... there's no mechanism to intercept synchronously unknown properties or methods invocation upfront in JS.

We have `__noSuchMethod__` from "legacy-land" but nothing like `__noSuchProperty__` in order to return `obj.something_like_the_following` as value.

**Comment by magcius** [comment]
Untested, but:

``` js
function camelCaseifyWord(word) {
    return word[0].toUpperCase() + word.slice(1);
}

function camelCaseifyName(name) {
    var parts = name.split('_');
    var first = parts[0], tail = parts.slice(1);
    return first + tail.map(camelCaseifyWord).join('');
}

function camelCaseify(obj) {
    Object.getOwnPropertyNames(obj).forEach(function(name) {
        var camelName = camelCaseifyName(name);
        var desc = Object.getOwnPropertyDescriptor(obj, name);
        Object.defineProperty(obj, camelName, desc);
    });
}

function camelCaseifyClass(constructor) {
    camelCaseify(constructor); // Static methods
    camelCaseify(constructor.prototype); // Instance methods / properties
}
```

**Comment by WebReflection** [comment]
1 that was my concern: does it work? Because it doesn't in GJS

2 `camelCase` really is _camelCase_, while `PascalCase` is _Capitalized_ You are doing what I've done already with a single regExp without needing all that map and split and stuff ... please let's keep in  mind less powerful and less RAM capable devices too (Raspberry Pi to name one) I work with many of them and abusing GC, Array and RAM via JS isn't exactly in my experience a good pattern (at least for a core functionality)

I might play with this too though, but I'll eventually create a module a part so it can be included via npm

**Comment by magcius** [comment]
It should work here because we don't rely on implicit property resolution like we do in gjs.

Also, sure, it was code for demonstration. I probably should have left the implementation blank.

**Comment by WebReflection** [comment]
FYI: I've created and published to npm [node-gtk-cc](https://github.com/WebReflection/node-gtk-cc) which is the camelCase version of this repository, based on branches and rewritten examples I've used here.

**Comment by magcius** [comment]
Er, why wasn't this a separate transform on top of `node-gtk` ?

**Comment by WebReflection** [comment]
Not sure I follow you. In the initial post you wrote

``` js
var ngc = require('node-gtk-camelcase');
```

as a way to obtain this camelcase version so I've created such version and published to npm, since it requires a different `package.json` file.

There is a branch called `camel-case` in here too but it got not much attention and it needs some merge conflict fix after [the PR regarding caching exported modules](https://github.com/WebReflection/node-gtk/pull/15).

I'm trying to speed up a little bit in parallel so we can test/experiment without conflicting at all.
Do you have any better/faster way to go about this camelCase idea?

**Comment by magcius** [comment]
But look at the line directly afterwards:

``` js
ngc.camelCaseifyAll(Gtk);
```

The _only_ thing you do with `ngc` is you modify a module imported with the original `node-gtk`. This module isn't a fork of the original `node-gtk` -- that's a maintenance nightmare! It just adds on new camel-case names for things.

**Comment by WebReflection** [comment]
If you require explicitly a camel case version of this project, what's the point of redundantly explicitly asking again to "camel-case-ify" anything?

I didn't want to have `package.json` conflicts within a branch in order to publish a dedicated module and I've imported from here because forking my own repo wasn't doing the right thing (or maybe I've done something wrong in there)

You don't have to maintain anything because that clone isn't meant to be maintained. We test, discuss, and write in here anything and I bring in there changes and publish so we can test these two versions without conflicting with examples, eventually tests, or anything else.

Does this make sense?

**Comment by magcius** [comment]
If you want camel casing to be implemented as part of the broader ecosystem, you're going to have to do it with a transformation, as I mentioned. It's not possible for the two libraries to both be imported as part of the same node instance.

I'm OK with your new project being a test module, but publishing it to npm seems a bit far to me in that case.

**Comment by WebReflection** [comment]
> It's not possible for the two libraries to both be imported as part of the same node instance.

Not sure if I'm doing what you mean but I've just installed both packages within a single instance without problems:

``` js
var
  ngtk = require('node-gtk'),
  ngtkcc = require('node-gtk-cc'),
  Gtk = ngtk.importNS('Gtk'),
  GtkCC = ngtkcc.importNS('Gtk')
;

ngtk.startLoop();
ngtkcc.startLoop();

Gtk.init(null);
GtkCC.init(null);

var win = new Gtk.Window({
  title: 'node-gtk',
  window_position: Gtk.WindowPosition.LEFT
});
win.set_default_size(200, 80);
win.add(new Gtk.Label({label: 'Hello Gtk+'}));

var wincc = new GtkCC.Window({
  title: 'node-gtk-cc',
  windowPosition: GtkCC.WindowPosition.CENTER
});
wincc.setDefaultSize(200, 80);
wincc.add(new GtkCC.Label({label: 'Hello Camel Gtk+'}));

win.show_all();
wincc.showAll();

Gtk.main();
GtkCC.main();
```

As a result I have two windows, one from `Gtk`, one from `GtkCC`.

Why shouldn't this be possible?

**Comment by magcius** [comment]
When C code needs to pass an instance of an object back to JS code, how does it know whether to use the camel case or the regular version? Currently, it's going to always pass a camel-case version because you simply overwrite the field in the GType field that tells it what the internal JS class is.

There's also complexities related to toggle refs as things get more complex. I wouldn't recommend it. Why are you so opposed to the camel-case transform, anyway?

**Comment by WebReflection** [comment]
I'm not, I'm just trying to understand what do you need/want.

If you want a different `require('package')` operation then I've no idea what you think would work or make sense. If you are asking for two different `importNS` and `importCCNS` then I know how to implement it but yet it'll be the exact same JS logic I've used right now.

You are talking about C code while everything I've done in order to make this project camelCase capable is on JS side so that absolutely nothing changes in your current backend logic and the problem you mention about C won't possibly happen.

Did you look at all at the `camel-case` branch or at how current camelCase logic/functionality has been implemented?

**Comment by magcius** [comment]
Here is what I want your package to export: `camelCaseifyAll`. It takes a namespace and then adds additional functions to it to camel-case it. It will traverse all classes and camel-case-ify those as well. It does not have any C code. It will not duplicate much existing code in `node-gtk`.

**Comment by WebReflection** [comment]
so you want both `Gtk.main_quit` and `Gtk.mainQuit` together after such operation plus the runtime pollution of all possible classes within the `Gtk` object?

This looks like an obtrusive approach to me ... not sure why you'd go for such solution.

I'll waste some time creating such repo but I am not sure its name will have anything to do with this project.

**Comment by magcius** [comment]
That's the only way it can work. Otherwise, if you use a JS library which uses `foo_bar`, it will break if you replace it with `fooBar`. You need both for existing code to keep working.

**Comment by WebReflection** [comment]
I've just shown you you can have both with two separate modules. Why you are still saying there's no other way is a mystery to me.

We don't deal with realm in here, you can have a camel case version of this project and that's what you use so your problematic scenario doesn't exist.

If another project uses non camel case version it's OK, these won't conflicts because these are two different modules, two different exports, different constructors, etcetera.

If I use in my project the CC version and I don't use CC properties or methods it's my fault.
Exposing both in a discoverable way is full of hidden footguns, IMO.

It ensures interoperability but that's exactly what you don't want to have: a VisualBasic like situation where either lower case or upper case properties access would have been the same.

I'll keep this `node-gtk-cc` up and running for testing purposes and also honestly I don't think we need a new repository for this very ad-hoc utility that only this project would probably ever require.

So, would an exported `importCCNS` work or a new export such `withCamelCase(name)` work? It seems more reasonable to me to have it here instead of elsewhere.

**Comment by WebReflection** [comment]
to provide more content:

``` js
var ngtk = require('node-gtk');
var Gtk = ngtk.importNS('Gtk');

// basic syntaxt
var GtkCC = ngtk.withCamelCase(Gtk);

// or direct simplified version with string overload
var GtkCC = ngtk.withCamelCase('Gtk');
```

So that if the argument is a string it asks for `importNS(str)` and then returns its polluted CC version, if needed (so even pollution, which is very expensive, should be cached).

In every case I need to have the cached module PR merged.

Thanks

**Comment by WebReflection** [comment]
Last immediate issue going forward this way, `Gtk.Window` and `Gtkcc.Window` will be two different functions pointing a the same prototype.

This is a major source of shenanigans when it comes to this weird interoperable scenario, unless we don't wrap by default everything and eventually pollute only if required.

This, on the other hand, will mutate namespaces at runtime when required and I'm not sure it's desired.

**Comment by magcius** [comment]
> I've just shown you you can have both with two separate modules.

Your example is a very simple example that popped up a window. As the project grows more complete, it will be impossible to use them side-by-side.

> If another project uses non camel case version it's OK, these won't conflicts because these are two different modules, two different exports, different constructors, etcetera.

Imagine a node module published which implements a cool widget I want in my app. Except that since they used the non-camel-case version and I used the camel-case-version, now I have to switch all my code over to the other module, since I can't add their widget to my code without changing over.

In my view, the absolute best solution would be to have absolutely one naming convention. I prefer the naming convention established by C / GTK+, you really prefer the naming convention established by JS. That's going to cause a rift right there. This bug is about finding technical ways of getting past that rift that allows a platform to be built on top.

**Comment by magcius** [comment]
> Last immediate issue going forward this way, Gtk.Window and Gtkcc.Window will be two different functions pointing a the same prototype.

Well, that means that `foo instanceof Gtk.Window` won't work. They need to be the same constructor. I know that constructors are difficult right now, and that's why I proposed #18, so we could patch constructors much easier.

**Comment by WebReflection** [comment]
Nope, they will both be `foo instanceof Gtk.Window` AND `foo instanceof GtkCC.Window` because like i've said they will point inevitably at the same prototype.

The constructor property of such prototype though, it cannot be both.

Having two different constructors is inevitable if these are augmented in JS because received objects need to be transformed as well.

Here the implementation to do that: https://github.com/WebReflection/node-gtk-cc/blob/master/lib/index.js#L203-L223

**Comment by magcius** [comment]
You could very well patch the constructor so it accepts both forms of arguments -- both camelCase and snake_case.

**Comment by WebReflection** [comment]
So this means you want a JS wrap by default, eventually enriched with camelCase-ability when `withCamelCase(namespace)` is invoked.

Is this correct?

**Comment by magcius** [comment]
What I was actually imagining was that we'd change constructors to simply call `this._init`, and your code would patch `Gtk.Window.prototype._init`, which wouldn't modify the constructor. But you can implement a separate constructor for now, since I haven't done the necessary refactoring for that yet.

**Comment by WebReflection** [comment]
is this `_init` thing already in place? If not, I strongly suggest to call it `__init__` to make it explicit is something nobody should change plus I'd define it non enumerable and non writable but configurable to avoid accidental rewrites too.

Please let me know when such `_init` or `__init__` lands, thanks.

**Comment by WebReflection** [comment]
in case you haven't received the notification ... mind reviewing this?
https://github.com/WebReflection/node-gtk/pull/19

It works already with the entire namespace (missing part is the **init** but if I understand correctly that's on you)

**Comment by WebReflection** [comment]
hey @magcius will you have time to check #19 ? I was planning to change node-gtk-cc in order to simply require this and overwrite its import to make it always camelCase but I need that PR to be merged.

If you have no time I'm going to merge it anyway, since it's a file a parte, and eventually we can discuss later what's missing or what should be changed.

Please let me know, thanks.

**Comment by WebReflection** [comment]
I have updated [this PR](https://github.com/WebReflection/node-gtk/pull/19) recently but I've seen zero activity by you @magcius ... is there anything else I can do to help in order to have the camelCase story completed?

Any possible ETA on your side?

Thanks

**Comment by magcius** [comment]
Sorry, I've been extremely busy with personal stuff this month and haven't had any, _any_ time for free time projects. I have been telling people internally at Endless enthusiastically about node-gtk, as well, so maybe we'll get more contributors too.

**Comment by WebReflection** [comment]
Well, thanks at least for finding time to reply. Looking forward for some update.

**Comment by WebReflection** [comment]
btw, congrats: https://www.gnome.org/news/2016/02/endless-joins-the-gnome-advisory-board/
